scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -816,6 +822,50 @@ sealed abstract class TypedPipe[+T] extends Serializable with Product {
   def toIterableExecution: Execution[Iterable[T]] =
     Execution.toIterable(this)
 
+  /**
+   * This gives an Execution that when run evaluates the TypedPipe,
+   * writes it to disk, and then gives you an Iterable that reads from
+   * disk on the submit node each time .iterator is called.
+   * Because of how scala Iterables work, mapping/flatMapping/filtering
+   * the Iterable forces a read of the entire thing. If you need it to
+   * be lazy, call .iterator and use the Iterator inside instead.
+   *
+   * Also this gives you Execution to clean up resources associated with Iterable.
+   *
+   * Example:
+   *
+   * pipe
+   *  .toIterableExecutionWithClose
+   *  .flatMap { case (iterable, close) =>
+   *    val distinctId = iterable.map(_.id).toSet
+   *
+   *    close.map(_ => distinctId)
+   *  }
+   *
+   * Note: prefer use `useAsIterator`
+   * Note: close will be no-op if typedPipe already forced on disk (i.e. you call forceToDisk before)
+   */
+  def toIterableExecutionWithClose: Execution[(Iterable[T], Execution[Unit])] =
+    Execution.toIterableWithClose(this)
+
+  /**
+   * When Execution run evaluates the TypedPipe, writes it to disk,
+   * and then gives you an Iterator that reads from disk on the submit node and clean up resources
+   * associated with Iterator after `fn` is done.
+   *
+   * WARNING: you must fully use the `Iterator` BEFORE you return Execution[B] in `fn`,
+   * because after `fn` is done we will clean up resources and data will be no longer available.
+   *
+   * Example:
+   *
+   * pipe.useAsIterator { iterator =>
+   *    val distinctId = iterable.map(_.id).toSet
+   *    Execution.from(distinctId)
+   * }
+   */
+  def useAsIterator[B](fn: Iterator[T] => Execution[B]): Execution[B] =
```

Comment:
the only thing I have some anxiety about is that we force people to only go through it once, whereas if we did `useAsIterable` they could go through it multiple times, e.g. sometimes even 2 passes is a big win: go through and find the max, then take all items > x% of the max. Doing that in one pass is a pain or maybe a lot of memory.